sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,502 @@
+                                                   PipelineOptions pipelineOptions) {
+      return serviceFactory.apply(pipelineOptions)
+          .splitIntoBundles(spec, desiredBundleSizeBytes, serviceFactory);
+    }
+  }
+
+  private static class CassandraReader<T> extends BoundedSource.BoundedReader<T> {
+
+    private final CassandraSource<T> source;
+    private final CassandraService<T> service;
+
+    private CassandraService.Reader<T> reader;
+
+    private CassandraReader(CassandraSource<T> source, CassandraService<T> service) {
+      this.source = source;
+      this.service = service;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      reader = service.createReader(source.splitQuery, source.spec.entity());
+      return reader.start();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      return reader.advance();
+    }
+
+    @Override
+    public void close() {
+      reader.close();
+      reader = null;
+    }
+
+    @Override
+    public BoundedSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public T getCurrent() {
+      return reader.getCurrent();
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      return Instant.now();
+    }
+  }
+
+  /**
+   * Estimates the size of a table using the Cassandra system table.
+   *
+   * @param connectionConfiguration The configuration of the connection to the cluster
+   * @param table The table to estimate
+   * @return The size of the table in bytes
+   */
+  public static long getEstimatedSize(
+      CassandraIO.ConnectionConfiguration connectionConfiguration, String table) {
+    try (Cluster cluster = connectionConfiguration.getCluster()) {
+      List<TokenRange> tokenRanges = getTokenRanges(cluster, connectionConfiguration, table);
+      long size = 0L;
+      for (TokenRange tokenRange : tokenRanges) {
+        size += tokenRange.meanPartitionSize * tokenRange.partitionCount;
+      }
+      return Math.round(size / getRingFraction(cluster, tokenRanges));
+    }
+  }
+
+  /**
+   * values that we get from system.size_estimates are for one node. We need to extrapolate to the
+   * whole cluster. This method estimates the percentage, the node represents in the cluster
+   *
+   * @param cluster The cassandra {@link Cluster} object
+   * @param tokenRanges The list of {@link TokenRange}
+   * @return The percentage the node represent in the whole cluster
+   */
+  //
+  private static float getRingFraction(Cluster cluster, List<TokenRange> tokenRanges) {
+    Partitioner partitioner = new Partitioner(cluster);
+    BigInteger addressedTokens = new BigInteger("0");
+    for (TokenRange tokenRange : tokenRanges) {
+      addressedTokens =
+          addressedTokens.add(distance(cluster, tokenRange.rangeStart, tokenRange.rangeEnd));
+    }
+    // it is < 1 because it is a percentage
+    return Float.valueOf(addressedTokens.divide(partitioner.getTotalTokens()).toString());
```

Comment:
Is this method tested? As written, it seems like it'll always return 0 because you're dividing an integer by a bigger integer.